src/repository/users.rs

```diff
@@ -0,0 +1,71 @@
+use axum::async_trait;
+use sqlx::PgPool;
+
+use crate::domain::users::{User, Credentials};
+
+pub enum UserGetError {
+    Missing,
+    Unknown
+}
+
+pub enum UserInsertError {
+    Unknown
+}
+
+#[async_trait]
+pub trait UserRepository {
+    async fn get_by_email(&self, email: &str) -> Result<User, UserGetError>;
+    async fn insert(&self, credentials: Credentials) -> Result<(), UserInsertError>;
+}
+
+#[derive(Debug, Clone)]
+pub struct PgUserRepository {
+    pub pool: PgPool
+}
+
+
+impl PgUserRepository {
+    pub fn new(pool: PgPool) -> Self {
+        Self { pool }
+    }
+}
+
+#[async_trait]
+impl UserRepository for PgUserRepository {
+    #[tracing::instrument]
+    async fn get_by_email(&self, email: &str) -> Result<User, UserGetError> {
+        let result = sqlx::query_as::<_, User>("SELECT * FROM users WHERE email = $1")
+            .bind(email)
+            .fetch_optional(&self.pool)
+            .await;
+        match result {
+            Ok(Some(user)) => Ok(user),
+            Ok(None) => {
+                tracing::warn!("User with email {} does not exist!", email);
+                Err(UserGetError::Missing)
+            },
+            Err(err) => {
+                tracing::error!(%err);
+                Err(UserGetError::Unknown)
+            }
```

Comment:
Use
```
use tracing::{error, warn};
```

Reply:
I don't know if it's good practice to include functional macros explicitly, but for logging it doesn't sound too bad, so I changed  `tracing::error` to `error` instead of vice versa.